### PR TITLE
add idea of foreign_id to enable easier langchain-vector store intera…

### PR DESCRIPTION
**Pull request by jeffchuber**
@dev2049 @hwchase17 - what do you all think about add a `foreign_id` to `Document`. 

See the test for Update included here is included to demonstrate why this is useful. 

It enables users to pass lists of `Documents` to vector store operations like `update`, `delete`, etc and handle the location of those records correctly. 

This is set up so that if the user sets the `foreign_id` before adding the records to the database, the database will use those `foreign_id`s as the ids. 

*** 

Concerns
- the name `foreign_id` implies that it is not something the user should set, where the user setting it is valid.
- if the user changes `foreign_id` after it has been passed to the database, the "link" will be lost. 
- I am not entirely familiar with the lifecycle of `Documents` inside LangChain and I don't know if this makes sense. 

The alternative to adding a `foreign_id` would be to match on the `page_content` - but I think this is far worse.

***

I couldn't get tests to run locally. Perhaps someone can point me in the right direction.

### CODE REVIEW

**Comment by dev2049** [comment]
thanks @jeffchuber! we've been thinking about what other keys to explicitly type on Documents. we'll probably add something like source and id/foreign_id soon but want to be really thoughtful about it. for now / for this pr could we keep foreign_id on metadata?

@eyurtsev any other thoughts?

File: langchain/vectorstores/chroma.py [hunk 1]
```diff
@@ -314,6 +320,72 @@ def delete_collection(self) -> None:
         """Delete the collection."""
         self._client.delete_collection(self._collection.name)
 
+    # should take document or list of documents
+    def update(self, documents: List[Document] | Document) -> None:
```
Comment (dev2049) [inline langchain/vectorstores/chroma.py:324]: nit: think `|` type hint operator only available in python 3.10+ and we still support 3.8+, will want to use Union

**Comment by jeffchuber** [comment]
@dev2049 that's interesting. 

In that case, I guess the cleanest way to do this would be to have something like `chroma_id` stuffed into the metadata whenever a `Document` object is passed to `Chroma` or created by `Chroma` (eg `add_texts`) -- and then stripped out before syncing back to Chroma (so the key-value doesn't get written to Chroma). This would mean that metadata in Chroma for an object would not match metadata in the Document, which I think is broadly a con b/c it is harder to reason about. 

The downside of this is that the attribute is ~strictly `read-only` in the sense that users can not set the property on the object, it can only be set for them. EG - it would be fairly obtuse for the user to set the foreign_id themselves in metadata and then have the expectation that to be respected in all operations.

`ids` in Chroma are currently immutable, though this isn't a hard and fast requirement of our API. 

This can also be solved by simply asking the user to maintain their own array of `Tuples` or other to keep track of `foreign_id-Document` pairs. In the long term, this feels like work that the user shouldn't have to do however. 

I totally respect that you are thinking more deeply about it! I mainly raised this PR to start/continue the conversation around this idea. I think I want to avoid workaround "magic" solutions for now, b/c that also will require that users migrate data down the road which is painful for everyone.

In any case - I'll keep PRs that do not require this change while you all consider how to approach it. And we can close this for now. (or keep it open for a few days to attract comments)

**Comment by eyurtsev** [comment]
A few points for discussion:

* Should documents be immutable? Making them immutable will likely help to prevent a bunch of issues and makes code easier to reason about. This complicates implementation of the pattern described in the PR.
* Documents could have their own uid assigned based on content hashes of the document. This does not require immutability, but makes more sense if immutability is present.
* What are administration patterns in production applications? Do users retain document objects after writing content into the search engine? Is it possible to phrase administration in terms of the `uid` associated with the documents, so no notion of foreign ids has to be introduced?
* What happens with foreign_id if a user needs to index the documents into multiple systems?

**Comment by eyurtsev** [comment]
@jeffchuber, we added the following indexing code support to LangChain. It should help manage indexing workflow for users. Let us know if you have any concerns

* https://github.com/langchain-ai/langchain/pull/9614
* https://python.langchain.com/docs/modules/data_connection/indexing


cc @baskaryan

**Comment by leo-gan** [comment]
@eyurtsev Does your last comment mean that we don't need this PR anymore? Is so, please, close this PR. TNX!
cc @jeffchuber

**Comment by efriis** [comment]
Closing because the PR wouldn't line up with the current directory structure of the library (would need to be in /libs/langchain/langchain instead of /langchain). Feel free to reopen against the current head if it's still relevant!

**Pull request closed without merging**